parachain/pallets/verifier-lightclient/src/lib.rs

```diff
@@ -197,17 +248,125 @@ impl<T: Trait> Module<T> {
 		// Maybe track new highest difficulty chain
 		let (_, highest_difficulty) = BestBlock::get();
 		if total_difficulty > highest_difficulty {
-			BestBlock::put((
-				EthereumHeaderId {
-					number: header.number,
-					hash,
-				},
-				total_difficulty,
-			));
+			let best_block_id = EthereumHeaderId {
+				number: header.number,
+				hash,
+			};
+			BestBlock::put((best_block_id, total_difficulty));
+
+			// Finalize blocks if possible
+			let finalized_block_id = FinalizedBlock::get();
+			let new_finalized_block_id = Self::get_best_finalized_header(
+				&best_block_id,
+				&finalized_block_id,
+			);
+			if new_finalized_block_id != finalized_block_id {
+				FinalizedBlock::put(new_finalized_block_id);
+			}
+	
+			// Clean up old headers
+			let pruning_range = BlocksToPrune::get();
+			let new_pruning_range = Self::prune_header_range(
+				&pruning_range,
+				HEADERS_TO_PRUNE_IN_SINGLE_IMPORT,
+				new_finalized_block_id.number.saturating_sub(FINALIZED_HEADERS_TO_KEEP),
+			);
+			if new_pruning_range != pruning_range {
+				BlocksToPrune::put(new_pruning_range);
+			}
 		}
 
 		Ok(())
 	}
+
+	// Return the latest block that can be finalized based on the given
+	// highest difficulty chain and previously finalized block.
+	fn get_best_finalized_header(
+		best_block_id: &EthereumHeaderId,
+		finalized_block_id: &EthereumHeaderId,
+	) -> EthereumHeaderId {
+		let mut new_finalized_block_id = finalized_block_id.clone();
+
+		let required_descendants = T::DescendantsUntilFinalized::get() as usize;
+		for (i, (ancestor_hash, ancestor)) in ancestry::<T>(best_block_id.hash).enumerate() {
+			if i < required_descendants {
+				continue;
+			}
+			if ancestor.number > finalized_block_id.number {
+				new_finalized_block_id = EthereumHeaderId {
+					hash: ancestor_hash,
+					number: ancestor.number,
+				};
+			} else {
+				assert!(ancestor_hash == finalized_block_id.hash);
```

Comment:
I'm still a bit uncomfortable with using asserts in production code. Even if theoretically there is no chance of a panic occuring. External reviewers/auditors may also raise concerns and ask about their implications.

As I demonstrated in the comment above, maybe raise a `Error::<T>::Undefined` in all cases where we would panic?

Reply:
This is only temporary while I test syncing with mainnet via the relayer. I'd prefer to create an issue to remove all assertions in a week or two. These assertions are saving me time right now. Thoughts?

Reply:
Ok, that makes sense, sounds good 👍🙂